File: src/domain/usecases/users/UpdateOwnThemeUseCase.ts
import type { User } from '@/domain/entities/User'
import type { UserRepository } from '@/domain/repositories/UserRepository'
import type { ThemePreference } from '@/domain/value-objects/ThemePreference'
import { isThemePreference } from '@/domain/value-objects/ThemePreference'
import {
  UnauthorizedError,
  ValidationError,
} from '@/domain/errors/DomainError'

export class UpdateOwnThemeUseCase {
  private readonly userRepository: UserRepository

  constructor(userRepository: UserRepository) {
    this.userRepository = userRepository
  }

  async execute(actor: User, theme: ThemePreference): Promise<User> {
    if (!actor.active) {
      throw new UnauthorizedError('Cuenta inactiva')
    }

    if (!isThemePreference(theme)) {
      throw new ValidationError('Tema inválido')
    }

    return this.userRepository.update(actor.id, { theme })
  }
}

File: consorcio/lib/domain/usecases/update_own_theme_use_case.dart
import '../entities/app_user.dart';
import '../errors/domain_exception.dart';
import '../repositories/user_repository.dart';
import '../value_objects/theme_preference.dart';

class UpdateOwnThemeUseCase {
  UpdateOwnThemeUseCase(this._userRepository);

  final UserRepository _userRepository;

  Future<AppUser> execute(AppUser actor, String theme) async {
    if (!actor.active) {
      throw DomainException('Cuenta inactiva');
    }

    final nextTheme = ThemePreference.normalize(theme);
    if (nextTheme != ThemePreference.light &&
        nextTheme != ThemePreference.dark) {
      throw DomainException('Tema inválido');
    }

    return _userRepository.updateTheme(
      userId: actor.id,
      theme: nextTheme,
    );
  }
}
